@joschi I am getting the following compilation error now:

```
[ERROR] graylog-project-repos/graylog-plugin-beats/src/main/java/org/graylog/plugins/beats/BeatsTransport.java:[45,9] constructor AbstractTcpTransport in class org.graylog2.plugin.inputs.transports.AbstractTcpTransport cannot be applied to given types;
  required: org.graylog2.plugin.configuration.Configuration,org.graylog2.plugin.inputs.util.ThroughputCounter,org.graylog2.plugin.LocalMetricRegistry,io.netty.channel.EventLoopGroup,org.graylog2.inputs.transports.netty.EventLoopGroupFactory,org.graylog2.inputs.transports.NettyTransportConfiguration
  found: org.graylog2.plugin.configuration.Configuration,org.graylog2.plugin.inputs.util.ThroughputCounter,org.graylog2.plugin.LocalMetricRegistry,io.netty.channel.EventLoopGroup,org.graylog2.inputs.transports.NettyTransportConfiguration
  reason: actual and formal argument lists differ in length
```

--- a/pom.xml
+++ b/pom.xml
@@ -200,6 +200,18 @@
                     <goals>package</goals>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <configuration>
+                    <systemProperties>
+                        <property>
+                            <name>io.netty.leakDetection.level</name>
+                            <value>PARANOID</value>
+                        </property>
+                    </systemProperties>
+                </configuration>
+            </plugin>
         </plugins>
     </build>
 </project>

--- a/src/main/java/org/graylog/plugins/beats/BeatsFrameDecoder.java
+++ b/src/main/java/org/graylog/plugins/beats/BeatsFrameDecoder.java
@@ -19,25 +19,24 @@ package org.graylog.plugins.beats;
 import com.fasterxml.jackson.core.JsonFactory;
 import com.fasterxml.jackson.core.JsonGenerator;
 import com.google.common.annotations.VisibleForTesting;
-import com.google.common.collect.Iterables;
 import com.google.common.io.ByteStreams;
 import com.google.common.primitives.Ints;
-import org.jboss.netty.buffer.ChannelBuffer;
-import org.jboss.netty.buffer.ChannelBuffers;
-import org.jboss.netty.channel.Channel;
-import org.jboss.netty.channel.ChannelHandlerContext;
-import org.jboss.netty.handler.codec.replay.ReplayingDecoder;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ReplayingDecoder;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import javax.annotation.Nullable;
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
-import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.List;
 import java.util.zip.InflaterInputStream;
 
@@ -68,26 +67,13 @@ public class BeatsFrameDecoder extends ReplayingDecoder<BeatsFrameDecoder.Decodi
     private long windowSize;
     private long sequenceNum;
 
-
     public BeatsFrameDecoder() {
-        super(DecodingState.PROTOCOL_VERSION, true);
+        super(DecodingState.PROTOCOL_VERSION);
     }
 
     @Override
-    protected Object decodeLast(ChannelHandlerContext ctx, Channel channel, ChannelBuffer buffer, DecodingState state) throws Exception {
-        // ignore, because can't send ACK after frame read
-        if (buffer.readable()) {
-            buffer.readBytes(super.actualReadableBytes());
-        }
-		        
-        checkpoint(DecodingState.PROTOCOL_VERSION);
-        return null;
-    }
-	
-    @Override
-    protected Object decode(ChannelHandlerContext ctx, Channel channel, ChannelBuffer buffer, DecodingState state) throws Exception {
-        ChannelBuffer[] events = null;
-        switch (state) {
+    protected void decode(ChannelHandlerContext channelHandlerContext, ByteBuf buffer, List<Object> list) throws Exception {
+        switch (state()) {
             case PROTOCOL_VERSION:
                 checkVersion(buffer);
                 checkpoint(DecodingState.FRAME_TYPE);
@@ -110,57 +96,49 @@ public class BeatsFrameDecoder extends ReplayingDecoder<BeatsFrameDecoder.Decodi
                     default:
                         throw new Exception("Unknown frame type: " + frameType);
                 }
-                return null;
+                return;
             case FRAME_WINDOW_SIZE:
                 processWindowSizeFrame(buffer);
                 break;
             case FRAME_DATA:
-                events = parseDataFrame(channel, buffer);
+                list.addAll(parseDataFrame(channelHandlerContext.channel(), buffer));
                 break;
             case FRAME_COMPRESSED:
-                events = processCompressedFrame(channel, buffer);
+                list.addAll(processCompressedFrame(channelHandlerContext.channel(), buffer));
                 break;
             case FRAME_JSON:
-                events = parseJsonFrame(channel, buffer);
+                list.addAll(parseJsonFrame(channelHandlerContext.channel(), buffer));
                 break;
             default:
-                throw new Exception("Unknown decoding state: " + state);
+                throw new Exception("Unknown decoding state: " + state());
         }
 
         checkpoint(DecodingState.PROTOCOL_VERSION);
-        return events;
     }
 
-    @Nullable
-    private ChannelBuffer[] processUncompressedBuffer(Channel channel, ChannelBuffer buffer) throws Exception {
+    private Collection<ByteBuf> processUncompressedBuffer(Channel channel, ByteBuf buffer) throws Exception {
         checkVersion(buffer);
         byte frameType = buffer.readByte();
 
-        ChannelBuffer[] events = null;
         switch (frameType) {
             case FRAME_WINDOW_SIZE:
                 processWindowSizeFrame(buffer);
-                break;
+                return Collections.emptyList();
             case FRAME_DATA:
-                events = parseDataFrame(channel, buffer);
-                break;
+                return parseDataFrame(channel, buffer);
             case FRAME_COMPRESSED:
-                events = processCompressedFrame(channel, buffer);
-                break;
+                return processCompressedFrame(channel, buffer);
             case FRAME_JSON:
-                events = parseJsonFrame(channel, buffer);
-                break;
+                return parseJsonFrame(channel, buffer);
             default:
                 throw new Exception("Unknown frame type: " + frameType);
         }
-
-        return events;
     }
 
-    private void checkVersion(ChannelBuffer channelBuffer) throws Exception {
+    private void checkVersion(ByteBuf channelBuffer) {
         byte version = channelBuffer.readByte();
         if (version != PROTOCOL_VERSION) {
-            throw new Exception("Unknown beats protocol version: " + version);
+            throw new IllegalStateException("Unknown beats protocol version: " + version);
         }
     }
 
@@ -169,60 +147,58 @@ public class BeatsFrameDecoder extends ReplayingDecoder<BeatsFrameDecoder.Decodi
      */
     private void sendACK(Channel channel) throws IOException {
         if (sequenceNum == windowSize) {
-            final ChannelBuffer buffer = ChannelBuffers.buffer(6);
+            final ByteBuf buffer = channel.alloc().buffer(6);
             buffer.writeByte(PROTOCOL_VERSION);
             buffer.writeByte(FRAME_ACK);
             buffer.writeInt((int) sequenceNum);
 
             LOG.trace("Sending ACK for sequence number {} on channel {}", sequenceNum, channel);
-            channel.write(buffer);
+            channel.writeAndFlush(buffer);
         }
     }
 
     /**
      * <a href="https://github.com/logstash-plugins/logstash-input-beats/blob/master/PROTOCOL.md#json-frame-type">'json' frame type</a>
      */
-    private ChannelBuffer[] parseJsonFrame(Channel channel, ChannelBuffer channelBuffer) throws IOException {
+    private Collection<ByteBuf> parseJsonFrame(Channel channel, ByteBuf channelBuffer) throws IOException {
         sequenceNum = channelBuffer.readUnsignedInt();
         LOG.trace("Received sequence number {}", sequenceNum);
 
         final int jsonLength = Ints.saturatedCast(channelBuffer.readUnsignedInt());
 
-        final ChannelBuffer buffer = channelBuffer.readSlice(jsonLength);
+        final ByteBuf buffer = channelBuffer.readBytes(jsonLength);
         sendACK(channel);
 
-        return new ChannelBuffer[]{buffer};
+        return Collections.singleton(buffer);
     }
 
     /**
      * @see <a href="https://github.com/logstash-plugins/logstash-input-beats/blob/master/PROTOCOL.md#compressed-frame-type">'compressed' frame type</a>
      */
-    private ChannelBuffer[] processCompressedFrame(Channel channel, ChannelBuffer channelBuffer) throws Exception {
+    private Collection<ByteBuf> processCompressedFrame(Channel channel, ByteBuf channelBuffer) throws Exception {
         final long payloadLength = channelBuffer.readUnsignedInt();
         final byte[] data = new byte[(int) payloadLength];
         channelBuffer.readBytes(data);
         try (final ByteArrayInputStream dataStream = new ByteArrayInputStream(data);
-            final InputStream in = new InflaterInputStream(dataStream)) {
-            final ChannelBuffer buffer = ChannelBuffers.wrappedBuffer(ByteStreams.toByteArray(in));
+             final InputStream in = new InflaterInputStream(dataStream)) {
+            final ByteBuf buffer = Unpooled.wrappedBuffer(ByteStreams.toByteArray(in));
             return processCompressedDataFrames(channel, buffer);
         }
     }
 
-    private ChannelBuffer[] processCompressedDataFrames(Channel channel, ChannelBuffer channelBuffer) throws Exception {
-        final List<ChannelBuffer> events = new ArrayList<>();
-        while (channelBuffer.readable()) {
-            final ChannelBuffer[] buffers = processUncompressedBuffer(channel, channelBuffer);
-            if (buffers != null) {
-                Iterables.addAll(events, Arrays.asList(buffers));
-            }
+    private Collection<ByteBuf> processCompressedDataFrames(Channel channel, ByteBuf channelBuffer) throws Exception {
+        final List<ByteBuf> events = new ArrayList<>();
+        while (channelBuffer.isReadable()) {
+            final Collection<ByteBuf> buffers = processUncompressedBuffer(channel, channelBuffer);
+            events.addAll(buffers);
         }
-        return events.toArray(new ChannelBuffer[0]);
+        return events;
     }
 
     /**
      * @see <a href="https://github.com/logstash-plugins/logstash-input-beats/blob/master/PROTOCOL.md#window-size-frame-type">'window size' frame type</a>
      */
-    private void processWindowSizeFrame(ChannelBuffer channelBuffer) {
+    private void processWindowSizeFrame(ByteBuf channelBuffer) {
         windowSize = channelBuffer.readUnsignedInt();
         LOG.trace("Changed window size to {}", windowSize);
     }
@@ -230,7 +206,7 @@ public class BeatsFrameDecoder extends ReplayingDecoder<BeatsFrameDecoder.Decodi
     /**
      * @see <a href="https://github.com/logstash-plugins/logstash-input-beats/blob/master/PROTOCOL.md#data-frame-type">'data' frame type</a>
      */
-    private ChannelBuffer[] parseDataFrame(Channel channel, ChannelBuffer channelBuffer) throws IOException {
+    private Collection<ByteBuf> parseDataFrame(Channel channel, ByteBuf channelBuffer) throws IOException {
         sequenceNum = channelBuffer.readUnsignedInt();
         LOG.trace("Received sequence number {}", sequenceNum);
 
@@ -247,17 +223,16 @@ public class BeatsFrameDecoder extends ReplayingDecoder<BeatsFrameDecoder.Decodi
             jg.writeEndObject();
         }
 
-        final ChannelBuffer buffer = ChannelBuffers.wrappedBuffer(outputStream.toByteArray());
+        final ByteBuf buffer = Unpooled.wrappedBuffer(outputStream.toByteArray());
         sendACK(channel);
 
-        return new ChannelBuffer[]{buffer};
+        return Collections.singleton(buffer);
     }
 
-    private String parseDataItem(ChannelBuffer channelBuffer) {
-        long length = channelBuffer.readUnsignedInt();
-        final byte[] bytes = new byte[(int) length];
-        channelBuffer.readBytes(bytes);
-        return new String(bytes, StandardCharsets.UTF_8);
+    private String parseDataItem(ByteBuf buf) {
+        int length = Ints.saturatedCast(buf.readUnsignedInt());
+        final ByteBuf item = buf.readSlice(length);
+        return item.toString(StandardCharsets.UTF_8);
     }
 
     @VisibleForTesting

--- a/src/main/java/org/graylog/plugins/beats/BeatsTransport.java
+++ b/src/main/java/org/graylog/plugins/beats/BeatsTransport.java
@@ -16,10 +16,11 @@
  */
 package org.graylog.plugins.beats;
 
-import com.codahale.metrics.InstrumentedExecutorService;
-import com.codahale.metrics.MetricRegistry;
-import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import com.google.inject.assistedinject.Assisted;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.EventLoopGroup;
+import org.graylog2.inputs.transports.NettyTransportConfiguration;
+import org.graylog2.inputs.transports.netty.EventLoopGroupFactory;
 import org.graylog2.plugin.LocalMetricRegistry;
 import org.graylog2.plugin.configuration.Configuration;
 import org.graylog2.plugin.configuration.ConfigurationRequest;
@@ -29,53 +30,27 @@ import org.graylog2.plugin.inputs.annotations.FactoryClass;
 import org.graylog2.plugin.inputs.transports.AbstractTcpTransport;
 import org.graylog2.plugin.inputs.transports.NettyTransport;
 import org.graylog2.plugin.inputs.transports.Transport;
-import org.graylog2.plugin.inputs.util.ConnectionCounter;
 import org.graylog2.plugin.inputs.util.ThroughputCounter;
-import org.jboss.netty.channel.ChannelHandler;
 
 import javax.inject.Inject;
-import javax.inject.Named;
 import java.util.LinkedHashMap;
 import java.util.concurrent.Callable;
-import java.util.concurrent.Executor;
-import java.util.concurrent.Executors;
-import java.util.concurrent.ThreadFactory;
-
-import static com.codahale.metrics.MetricRegistry.name;
 
 public class BeatsTransport extends AbstractTcpTransport {
     @Inject
     public BeatsTransport(@Assisted Configuration configuration,
+                          EventLoopGroup eventLoopGroup,
+                          EventLoopGroupFactory eventLoopGroupFactory,
+                          NettyTransportConfiguration nettyTransportConfiguration,
                           ThroughputCounter throughputCounter,
-                          LocalMetricRegistry localRegistry,
-                          @Named("bossPool") Executor bossPool,
-                          ConnectionCounter connectionCounter) {
-        this(configuration, throughputCounter, localRegistry, bossPool, executorService("beats-worker", "beats-transport-worker-%d", localRegistry), connectionCounter);
-    }
-
-    private BeatsTransport(Configuration configuration,
-                           ThroughputCounter throughputCounter,
-                           LocalMetricRegistry localRegistry,
-                           Executor bossPool,
-                           Executor workerPool,
-                           ConnectionCounter connectionCounter) {
-        super(configuration, throughputCounter, localRegistry, bossPool, workerPool, connectionCounter);
-    }
-
-    private static Executor executorService(final String executorName, final String threadNameFormat, final MetricRegistry metricRegistry) {
-        final ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat(threadNameFormat).build();
-        return new InstrumentedExecutorService(
-                Executors.newCachedThreadPool(threadFactory),
-                metricRegistry,
-                name(BeatsTransport.class, executorName, "executor-service"));
+                          LocalMetricRegistry localRegistry) {
+        super(configuration, throughputCounter, localRegistry, eventLoopGroup, eventLoopGroupFactory, nettyTransportConfiguration);
     }
 
     @Override
-    protected LinkedHashMap<String, Callable<? extends ChannelHandler>> getFinalChannelHandlers(MessageInput input) {
-        final LinkedHashMap<String, Callable<? extends ChannelHandler>> finalChannelHandlers = super.getFinalChannelHandlers(input);
-        final LinkedHashMap<String, Callable<? extends ChannelHandler>> handlers = new LinkedHashMap<>();
+    protected LinkedHashMap<String, Callable<? extends ChannelHandler>> getCustomChildChannelHandlers(MessageInput input) {
+        final LinkedHashMap<String, Callable<? extends ChannelHandler>> handlers = new LinkedHashMap<>(super.getCustomChildChannelHandlers(input));
         handlers.put("beats", BeatsFrameDecoder::new);
-        handlers.putAll(finalChannelHandlers);
 
         return handlers;
     }

--- a/src/test/java/org/graylog/plugins/beats/BeatsFrameDecoderTest.java
+++ b/src/test/java/org/graylog/plugins/beats/BeatsFrameDecoderTest.java
@@ -19,10 +19,11 @@ package org.graylog.plugins.beats;
 import com.fasterxml.jackson.core.type.TypeReference;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.google.common.collect.ImmutableMap;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.embedded.EmbeddedChannel;
+import io.netty.handler.logging.LoggingHandler;
 import org.graylog2.shared.bindings.providers.ObjectMapperProvider;
-import org.jboss.netty.buffer.ChannelBuffer;
-import org.jboss.netty.buffer.ChannelBuffers;
-import org.jboss.netty.handler.codec.embedder.DecoderEmbedder;
 import org.junit.Before;
 import org.junit.Test;
 
@@ -34,32 +35,31 @@ import static org.assertj.core.api.Assertions.assertThat;
 
 public class BeatsFrameDecoderTest {
     private final ObjectMapper objectMapper = new ObjectMapperProvider().get();
-    private DecoderEmbedder<Iterable<ChannelBuffer>> channel;
+    private EmbeddedChannel channel;
     private BeatsFrameDecoder decoder;
 
     @Before
     public void setUp() throws Exception {
         decoder = new BeatsFrameDecoder();
-        channel = new DecoderEmbedder<>(decoder);
+        channel = new EmbeddedChannel(new LoggingHandler(), decoder);
     }
 
     @Test
     public void decodeWindowSizeFrame() throws Exception {
-        final ChannelBuffer buffer = buildWindowSizeFrame(1234);
+        final ByteBuf buffer = buildWindowSizeFrame(1234);
 
-        while (buffer.readable()) {
-            channel.offer(buffer);
+        while (buffer.isReadable()) {
+            channel.writeInbound(buffer);
         }
         channel.finish();
         assertThat(buffer.readableBytes()).isEqualTo(0);
 
-        final ChannelBuffer[] output = channel.pollAll(new ChannelBuffer[0]);
-        assertThat(output).isEmpty();
+        assertThat((Object) channel.readInbound()).isNull();
         assertThat(decoder.getWindowSize()).isEqualTo(1234L);
     }
 
-    private ChannelBuffer buildWindowSizeFrame(int windowSize) {
-        final ChannelBuffer buffer = ChannelBuffers.buffer(6);
+    private ByteBuf buildWindowSizeFrame(int windowSize) {
+        final ByteBuf buffer = Unpooled.buffer(6);
         buffer.writeByte('2');
         buffer.writeByte('W');
         buffer.writeInt(windowSize);
@@ -71,22 +71,19 @@ public class BeatsFrameDecoderTest {
         final String json = "{\"answer\": 42}";
         final byte[] jsonBytes = json.getBytes(StandardCharsets.UTF_8);
 
-        final ChannelBuffer buffer = buildJsonFrame(jsonBytes, 0);
+        final ByteBuf buffer = buildJsonFrame(jsonBytes, 0);
 
-        while (buffer.readable()) {
-            channel.offer(buffer);
+        while (buffer.isReadable()) {
+            channel.writeInbound(buffer);
         }
         channel.finish();
 
         assertThat(buffer.readableBytes()).isEqualTo(0);
 
-        final ChannelBuffer[] output = channel.pollAll(new ChannelBuffer[0]);
-        assertThat(output).hasSize(2);
-
-        final ChannelBuffer replyBuffer = output[0];
+        final ByteBuf replyBuffer = channel.readOutbound();
         assertThat(extractSequenceNumber(replyBuffer)).isEqualTo(0L);
 
-        final ChannelBuffer resultBuffer = output[1];
+        final ByteBuf resultBuffer = channel.readInbound();
         final byte[] resultBytes = new byte[resultBuffer.readableBytes()];
         resultBuffer.readBytes(resultBytes);
         final Map<String, Object> result = objectMapper.readValue(resultBytes, new TypeReference<Map<String, Object>>() {
@@ -96,8 +93,8 @@ public class BeatsFrameDecoderTest {
                 .containsEntry("answer", 42);
     }
 
-    private ChannelBuffer buildJsonFrame(byte[] jsonBytes, int sequenceNum) {
-        final ChannelBuffer buffer = ChannelBuffers.buffer(10 + jsonBytes.length);
+    private ByteBuf buildJsonFrame(byte[] jsonBytes, int sequenceNum) {
+        final ByteBuf buffer = Unpooled.buffer(10 + jsonBytes.length);
         buffer.writeByte('2');
         buffer.writeByte('J');
         // Sequence number
@@ -115,22 +112,19 @@ public class BeatsFrameDecoderTest {
                 "foo", "bar",
                 "quux", "baz");
 
-        final ChannelBuffer buffer = buildDataFrame(data, 0);
+        final ByteBuf buffer = buildDataFrame(data, 0);
 
-        while (buffer.readable()) {
-            channel.offer(buffer);
+        while (buffer.isReadable()) {
+            channel.writeInbound(buffer);
         }
         channel.finish();
 
         assertThat(buffer.readableBytes()).isEqualTo(0);
 
-        final ChannelBuffer[] output = channel.pollAll(new ChannelBuffer[0]);
-        assertThat(output).hasSize(2);
-
-        final ChannelBuffer replyBuffer = output[0];
+        final ByteBuf replyBuffer = channel.readOutbound();
         assertThat(extractSequenceNumber(replyBuffer)).isEqualTo(0L);
 
-        final ChannelBuffer resultBuffer = output[1];
+        final ByteBuf resultBuffer = channel.readInbound();
         final byte[] resultBytes = new byte[resultBuffer.readableBytes()];
         resultBuffer.readBytes(resultBytes);
         final Map<String, Object> result = objectMapper.readValue(resultBytes, new TypeReference<Map<String, Object>>() {
@@ -138,8 +132,8 @@ public class BeatsFrameDecoderTest {
         assertThat(result).isEqualTo(data);
     }
 
-    private ChannelBuffer buildDataFrame(Map<String, String> data, int sequenceNum) {
-        final ChannelBuffer buffer = ChannelBuffers.dynamicBuffer();
+    private ByteBuf buildDataFrame(Map<String, String> data, int sequenceNum) {
+        final ByteBuf buffer = Unpooled.buffer();
         buffer.writeByte('2');
         buffer.writeByte('D');
         // Sequence number
@@ -163,23 +157,20 @@ public class BeatsFrameDecoderTest {
         final String json = "{\"answer\": 42}";
         final byte[] jsonBytes = json.getBytes(StandardCharsets.UTF_8);
 
-        final ChannelBuffer innerBuffer = buildJsonFrame(jsonBytes, 0);
-        final ChannelBuffer buffer = buildCompressedFrame(innerBuffer.array(), 3);
+        final ByteBuf innerBuffer = buildJsonFrame(jsonBytes, 0);
+        final ByteBuf buffer = buildCompressedFrame(innerBuffer.array(), 3);
 
-        while (buffer.readable()) {
-            channel.offer(buffer);
+        while (buffer.isReadable()) {
+            channel.writeInbound(buffer);
         }
         channel.finish();
 
         assertThat(buffer.readableBytes()).isEqualTo(0);
 
-        final ChannelBuffer[] output = channel.pollAll(new ChannelBuffer[0]);
-        assertThat(output).hasSize(2);
-
-        final ChannelBuffer replyBuffer = output[0];
+        final ByteBuf replyBuffer = channel.readOutbound();
         assertThat(extractSequenceNumber(replyBuffer)).isEqualTo(0L);
 
-        final ChannelBuffer resultBuffer = output[1];
+        final ByteBuf resultBuffer = channel.readInbound();
         final byte[] resultBytes = new byte[resultBuffer.readableBytes()];
         resultBuffer.readBytes(resultBytes);
         final Map<String, Object> result = objectMapper.readValue(resultBytes, new TypeReference<Map<String, Object>>() {
@@ -189,7 +180,7 @@ public class BeatsFrameDecoderTest {
                 .containsEntry("answer", 42);
     }
 
-    private ChannelBuffer buildCompressedFrame(byte[] payload, int compressionLevel) {
+    private ByteBuf buildCompressedFrame(byte[] payload, int compressionLevel) {
         final Deflater deflater = new Deflater(compressionLevel);
         deflater.setInput(payload);
         deflater.finish();
@@ -198,7 +189,7 @@ public class BeatsFrameDecoderTest {
         final int compressedPayloadLength = deflater.deflate(compressedPayload);
         deflater.end();
 
-        final ChannelBuffer buffer = ChannelBuffers.buffer(6 + compressedPayloadLength);
+        final ByteBuf buffer = Unpooled.buffer(6 + compressedPayloadLength);
         buffer.writeByte('2');
         buffer.writeByte('C');
         // Compressed payload length
@@ -211,30 +202,31 @@ public class BeatsFrameDecoderTest {
     @Test
     public void decodeMultipleFrames() throws Exception {
         final Map<String, String> data = ImmutableMap.of("foo", "bar");
-        final ChannelBuffer buffer = ChannelBuffers.copiedBuffer(
+        final ByteBuf buffer = Unpooled.copiedBuffer(
                 buildWindowSizeFrame(2),
                 buildDataFrame(data, 0),
                 buildDataFrame(data, 1),
                 buildDataFrame(data, 2)
         );
 
-        while (buffer.readable()) {
-            channel.offer(buffer);
+        while (buffer.isReadable()) {
+            channel.writeInbound(buffer);
         }
         channel.finish();
 
         assertThat(buffer.readableBytes()).isEqualTo(0);
 
-        final ChannelBuffer[] output = channel.pollAll(new ChannelBuffer[0]);
-        assertThat(output).hasSize(4);
+        final ByteBuf output1 = channel.readInbound();
+        final ByteBuf output2 = channel.readInbound();
+        final ByteBuf output3 = channel.readInbound();
         assertThat(decoder.getWindowSize()).isEqualTo(2);
         assertThat(decoder.getSequenceNum()).isEqualTo(2L);
 
-        final ChannelBuffer replyBuffer = output[2];
+        final ByteBuf replyBuffer = channel.readOutbound();
         assertThat(extractSequenceNumber(replyBuffer)).isEqualTo(2L);
 
-        final ChannelBuffer[] received = {output[0], output[1], output[3]};
-        for (ChannelBuffer resultBuffer : received) {
+        final ByteBuf[] received = {output1, output2, output3};
+        for (ByteBuf resultBuffer : received) {
             final byte[] resultBytes = new byte[resultBuffer.readableBytes()];
             resultBuffer.readBytes(resultBytes);
             final Map<String, Object> result = objectMapper.readValue(resultBytes, new TypeReference<Map<String, Object>>() {
@@ -243,7 +235,7 @@ public class BeatsFrameDecoderTest {
         }
     }
 
-    private long extractSequenceNumber(ChannelBuffer buffer) {
+    private long extractSequenceNumber(ByteBuf buffer) {
         assertThat(buffer.readByte()).isEqualTo((byte) '2');
         assertThat(buffer.readByte()).isEqualTo((byte) 'A');
         final long seqNum = buffer.readUnsignedInt();

--- a/src/test/java/org/graylog/plugins/beats/BeatsTransportTest.java
+++ b/src/test/java/org/graylog/plugins/beats/BeatsTransportTest.java
@@ -16,35 +16,47 @@
  */
 package org.graylog.plugins.beats;
 
+import io.netty.channel.nio.NioEventLoopGroup;
+import org.graylog2.inputs.transports.NettyTransportConfiguration;
+import org.graylog2.inputs.transports.netty.EventLoopGroupFactory;
 import org.graylog2.plugin.LocalMetricRegistry;
 import org.graylog2.plugin.configuration.Configuration;
 import org.graylog2.plugin.inputs.MessageInput;
-import org.graylog2.plugin.inputs.util.ConnectionCounter;
 import org.graylog2.plugin.inputs.util.ThroughputCounter;
-import org.jboss.netty.channel.ChannelHandler;
-import org.jboss.netty.util.HashedWheelTimer;
+import org.junit.After;
+import org.junit.Before;
 import org.junit.Test;
-
-import java.util.LinkedHashMap;
-import java.util.concurrent.Callable;
-import java.util.concurrent.Executors;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.Mockito.mock;
 
 public class BeatsTransportTest {
+    private NioEventLoopGroup eventLoopGroup;
+
+    @Before
+    public void setUp() {
+        eventLoopGroup = new NioEventLoopGroup(1);
+    }
+
+    @After
+    public void tearDown() {
+        eventLoopGroup.shutdownGracefully();
+    }
+
     @Test
-    public void getFinalChannelHandlers() throws Exception {
+    public void customChildChannelHandlersContainBeatsHandler() {
+        final NettyTransportConfiguration nettyTransportConfiguration = new NettyTransportConfiguration("nio", "jdk", 1);
+        final EventLoopGroupFactory eventLoopGroupFactory = new EventLoopGroupFactory(nettyTransportConfiguration);
         final BeatsTransport transport = new BeatsTransport(
-                new Configuration(null),
-                new ThroughputCounter(new HashedWheelTimer()),
-                new LocalMetricRegistry(),
-                Executors.newSingleThreadExecutor(),
-                new ConnectionCounter()
+                Configuration.EMPTY_CONFIGURATION,
+                eventLoopGroup,
+                eventLoopGroupFactory,
+                nettyTransportConfiguration,
+                new ThroughputCounter(eventLoopGroup),
+                new LocalMetricRegistry()
         );
 
         final MessageInput input = mock(MessageInput.class);
-        final LinkedHashMap<String, Callable<? extends ChannelHandler>> channelHandlers = transport.getFinalChannelHandlers(input);
-        assertThat(channelHandlers).containsKey("beats");
+        assertThat(transport.getCustomChildChannelHandlers(input)).containsKey("beats");
     }
 }

--- a/src/test/java/org/graylog/plugins/beats/ConsolePrinter.java
+++ b/src/test/java/org/graylog/plugins/beats/ConsolePrinter.java
@@ -18,28 +18,26 @@ package org.graylog.plugins.beats;
 
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.google.common.primitives.Ints;
+import io.netty.bootstrap.ServerBootstrap;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.handler.logging.LoggingHandler;
 import org.graylog2.plugin.Message;
 import org.graylog2.plugin.configuration.Configuration;
 import org.graylog2.plugin.journal.RawMessage;
 import org.graylog2.shared.bindings.providers.ObjectMapperProvider;
-import org.jboss.netty.bootstrap.ServerBootstrap;
-import org.jboss.netty.buffer.ChannelBuffer;
-import org.jboss.netty.channel.Channel;
-import org.jboss.netty.channel.ChannelFactory;
-import org.jboss.netty.channel.ChannelHandlerContext;
-import org.jboss.netty.channel.ExceptionEvent;
-import org.jboss.netty.channel.MessageEvent;
-import org.jboss.netty.channel.SimpleChannelUpstreamHandler;
-import org.jboss.netty.channel.socket.nio.NioServerSocketChannelFactory;
-import org.jboss.netty.handler.logging.LoggingHandler;
-
-import java.net.InetSocketAddress;
-import java.util.concurrent.Executors;
 
 import static com.google.common.base.MoreObjects.firstNonNull;
 
 public class ConsolePrinter {
-    public static void main(String[] args) {
+    public static void main(String[] args) throws Exception {
         String hostname = "127.0.0.1";
         int port = 5044;
         if (args.length >= 2) {
@@ -50,25 +48,35 @@ public class ConsolePrinter {
             port = firstNonNull(Ints.tryParse(args[1]), 5044);
         }
 
-        final ChannelFactory factory =
-                new NioServerSocketChannelFactory(
-                        Executors.newCachedThreadPool(),
-                        Executors.newCachedThreadPool());
-        final ServerBootstrap b = new ServerBootstrap(factory);
-        b.getPipeline().addLast("beats-frame-decoder", new BeatsFrameDecoder());
-        b.getPipeline().addLast("beats-codec", new BeatsCodecHandler());
-        b.getPipeline().addLast("logging", new LoggingHandler());
-        System.out.println("Starting listener on " + hostname + ":" + port);
-        b.bind(new InetSocketAddress(hostname, port));
+
+        final EventLoopGroup eventLoopGroup = new NioEventLoopGroup();
+        try {
+            final ServerBootstrap b = new ServerBootstrap()
+                    .group(eventLoopGroup)
+                    .channel(NioServerSocketChannel.class)
+                    .childHandler(new ChannelInitializer<SocketChannel>() {
+                        @Override
+                        public void initChannel(SocketChannel ch) throws Exception {
+                            ch.pipeline().addLast("logging", new LoggingHandler());
+                            ch.pipeline().addLast("beats-frame-decoder", new BeatsFrameDecoder());
+                            ch.pipeline().addLast("beats-codec", new BeatsCodecHandler());
+                        }
+                    });
+
+            System.out.println("Starting listener on " + hostname + ":" + port);
+            final ChannelFuture future = b.bind(hostname, port).sync();
+            future.channel().closeFuture().sync();
+        } finally {
+            eventLoopGroup.shutdownGracefully();
+        }
     }
 
-    public static class BeatsCodecHandler extends SimpleChannelUpstreamHandler {
+    public static class BeatsCodecHandler extends SimpleChannelInboundHandler<ByteBuf> {
         private final ObjectMapper objectMapper = new ObjectMapperProvider().get();
         private final BeatsCodec beatsCodec = new BeatsCodec(Configuration.EMPTY_CONFIGURATION, objectMapper);
 
         @Override
-        public void messageReceived(ChannelHandlerContext ctx, MessageEvent e) throws Exception {
-            final ChannelBuffer message = (ChannelBuffer) e.getMessage();
+        protected void channelRead0(ChannelHandlerContext ctx, ByteBuf message) throws Exception {
             final int readableBytes = message.readableBytes();
             final byte[] messageBytes = new byte[readableBytes];
             message.readBytes(messageBytes);
@@ -77,15 +85,7 @@ public class ConsolePrinter {
             final Message decodedMessage = beatsCodec.decode(rawMessage);
             System.out.println(decodedMessage);
 
-            super.messageReceived(ctx, e);
-        }
-
-        @Override
-        public void exceptionCaught(ChannelHandlerContext ctx, ExceptionEvent e) throws Exception {
-            e.getCause().printStackTrace();
-
-            final Channel ch = e.getChannel();
-            ch.close();
+            ctx.fireChannelRead(decodedMessage);
         }
     }
 }

